### PR TITLE
5695: Fix docker build issue

### PR DESCRIPTION
**Pull request by sam-roberts-bjss**

### CODE REVIEW

File: src/NHSD.BuyingCatalogue.Identity.Api/Dockerfile [hunk 1]
```diff
@@ -1,29 +1,39 @@
...
+COPY "src/NHSD.BuyingCatalogue.Identity.Common/NHSD.BuyingCatalogue.Identity.Common.csproj" "src/NHSD.BuyingCatalogue.Identity.Common/NHSD.BuyingCatalogue.Identity.Common.csproj"
+COPY "src/NHSD.BuyingCatalogue.Identity.Api/NHSD.BuyingCatalogue.Identity.Api.csproj" "src/NHSD.BuyingCatalogue.Identity.Api/NHSD.BuyingCatalogue.Identity.Api.csproj"
+COPY "tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests.csproj" "tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests.csproj"
+COPY "tests/NHSD.BuyingCatalogue.Organisations.Api.UnitTests/NHSD.BuyingCatalogue.Organisations.Api.UnitTests.csproj" "tests/NHSD.BuyingCatalogue.Organisations.Api.UnitTests/NHSD.BuyingCatalogue.Organisations.Api.UnitTests.csproj"
+COPY "tests/NHSD.BuyingCatalogue.Identity.Api.Testing.Data/NHSD.BuyingCatalogue.Identity.Api.Testing.Data.csproj" "tests/NHSD.BuyingCatalogue.Identity.Api.Testing.Data/NHSD.BuyingCatalogue.Identity.Api.Testing.Data.csproj"
```
Comment (assoomro) [inline src/NHSD.BuyingCatalogue.Identity.Api/Dockerfile]: Same as before

File: samples/NHSD.BuyingCatalogue.Identity.Api.SampleResource/Dockerfile [hunk 2]
```diff
@@ -1,20 +1,31 @@
...
+
 COPY . .
-WORKDIR "/src/samples/NHSD.BuyingCatalogue.Identity.Api.SampleResource"
-RUN dotnet build "NHSD.BuyingCatalogue.Identity.Api.SampleResource.csproj" -c Release -o /app/build
+WORKDIR "/samples/NHSD.BuyingCatalogue.Identity.Api.SampleResource"
+RUN dotnet publish --no-restore -c Release -o /app
 
-FROM build AS publish
-RUN dotnet publish "NHSD.BuyingCatalogue.Identity.Api.SampleResource.csproj" -c Release -o /app/publish
+FROM build as publish
```
Comment (assoomro) [inline samples/NHSD.BuyingCatalogue.Identity.Api.SampleResource/Dockerfile]: This is an empty stage I believe. I'd either
a) get rid of it and rename the copy statement in the final stage
b) move the dotnet publish step here

File: samples/NHSD.BuyingCatalogue.Identity.Api.SampleMvcClient/Dockerfile [hunk 3]
```diff
@@ -1,22 +1,31 @@
...
+
 COPY . .
-WORKDIR "/src/samples/NHSD.BuyingCatalogue.Identity.Api.SampleMvcClient"
-RUN dotnet build "NHSD.BuyingCatalogue.Identity.Api.SampleMvcClient.csproj" -c Release -o /app/build
+WORKDIR "/samples/NHSD.BuyingCatalogue.Identity.Api.SampleMvcClient"
+RUN dotnet publish --no-restore -c Release -o /app
 
-FROM build AS publish
-RUN dotnet publish "NHSD.BuyingCatalogue.Identity.Api.SampleMvcClient.csproj" -c Release -o /app/publish
+FROM build as publish
```
Comment (assoomro) [inline samples/NHSD.BuyingCatalogue.Identity.Api.SampleMvcClient/Dockerfile]: This is an empty stage I believe. I'd either
a) get rid of it and rename the copy statement in the final stage
b) move the dotnet publish step here

File: samples/NHSD.BuyingCatalogue.Identity.Api.SampleResource/Dockerfile [hunk 4]
```diff
@@ -1,20 +1,31 @@
 FROM mcr.microsoft.com/dotnet/core/aspnet:3.1-buster-slim AS base
 WORKDIR /app
 EXPOSE 80
+EXPOSE 443
```
Comment (assoomro) [inline samples/NHSD.BuyingCatalogue.Identity.Api.SampleResource/Dockerfile]: Same question as before.

File: src/NHSD.BuyingCatalogue.Identity.Api/Dockerfile [hunk 5]
```diff
@@ -1,29 +1,39 @@
...
+COPY --from=node-build /node/wwwroot/css wwwroot/css
 WORKDIR "/src/NHSD.BuyingCatalogue.Identity.Api"
+RUN dotnet publish --no-restore -c Release -o /app
 
-COPY --from=pre-build /node/wwwroot/css wwwroot/css
-RUN dotnet build "NHSD.BuyingCatalogue.Identity.Api.csproj" -c Release -o /app/build
-
-FROM build AS publish
-RUN dotnet publish "NHSD.BuyingCatalogue.Identity.Api.csproj" -c Release -o /app/publish
+FROM build as publish
```
Comment (assoomro) [inline src/NHSD.BuyingCatalogue.Identity.Api/Dockerfile]: Same as above

File: src/NHSD.BuyingCatalogue.Organisations.Api/Dockerfile [hunk 6]
```diff
@@ -1,22 +1,31 @@
...
+COPY "src/NHSD.BuyingCatalogue.Identity.Common/NHSD.BuyingCatalogue.Identity.Common.csproj" "src/NHSD.BuyingCatalogue.Identity.Common/NHSD.BuyingCatalogue.Identity.Common.csproj"
+COPY "src/NHSD.BuyingCatalogue.Identity.Api/NHSD.BuyingCatalogue.Identity.Api.csproj" "src/NHSD.BuyingCatalogue.Identity.Api/NHSD.BuyingCatalogue.Identity.Api.csproj"
+COPY "tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests.csproj" "tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests.csproj"
+COPY "tests/NHSD.BuyingCatalogue.Organisations.Api.UnitTests/NHSD.BuyingCatalogue.Organisations.Api.UnitTests.csproj" "tests/NHSD.BuyingCatalogue.Organisations.Api.UnitTests/NHSD.BuyingCatalogue.Organisations.Api.UnitTests.csproj"
+COPY "tests/NHSD.BuyingCatalogue.Identity.Api.Testing.Data/NHSD.BuyingCatalogue.Identity.Api.Testing.Data.csproj" "tests/NHSD.BuyingCatalogue.Identity.Api.Testing.Data/NHSD.BuyingCatalogue.Identity.Api.Testing.Data.csproj"
```
Comment (assoomro) [inline src/NHSD.BuyingCatalogue.Organisations.Api/Dockerfile]: Same as above

File: src/NHSD.BuyingCatalogue.Organisations.Api/Dockerfile [hunk 7]
```diff
@@ -1,22 +1,31 @@
...
+RUN dotnet restore "NHSD.BuyingCatalogue.Identity-ExludeDatabase.sln"
 
+COPY . .
 WORKDIR "/src/NHSD.BuyingCatalogue.Organisations.Api"
-RUN dotnet build "NHSD.BuyingCatalogue.Organisations.Api.csproj" -c Release -o /app/build
+RUN dotnet publish --no-restore -c Release -o /app
 
-FROM build AS publish
-RUN dotnet publish "NHSD.BuyingCatalogue.Organisations.Api.csproj" -c Release -o /app/publish
+FROM build as publish
```
Comment (assoomro) [inline src/NHSD.BuyingCatalogue.Organisations.Api/Dockerfile]: Same as above

File: samples/NHSD.BuyingCatalogue.Identity.Api.SampleMvcClient/Dockerfile [hunk 8]
```diff
@@ -1,22 +1,31 @@
-#See https://aka.ms/containerfastmode to understand how Visual Studio uses this Dockerfile to build your images for faster debugging.
-
 FROM mcr.microsoft.com/dotnet/core/aspnet:3.1-buster-slim AS base
 WORKDIR /app
 EXPOSE 80
+EXPOSE 443
```
[the remark this answers is not in the document]
Comment (sam-roberts-bjss) [reply samples/NHSD.BuyingCatalogue.Identity.Api.SampleMvcClient/Dockerfile]: We may want this in the future but as of now, no we don't use it. Would you like me to remove it?

File: samples/NHSD.BuyingCatalogue.Identity.Api.SampleMvcClient/Dockerfile [hunk 9]
```diff
@@ -1,22 +1,31 @@
...
+COPY "src/NHSD.BuyingCatalogue.Identity.Common/NHSD.BuyingCatalogue.Identity.Common.csproj" "src/NHSD.BuyingCatalogue.Identity.Common/NHSD.BuyingCatalogue.Identity.Common.csproj"
+COPY "src/NHSD.BuyingCatalogue.Identity.Api/NHSD.BuyingCatalogue.Identity.Api.csproj" "src/NHSD.BuyingCatalogue.Identity.Api/NHSD.BuyingCatalogue.Identity.Api.csproj"
+COPY "tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests.csproj" "tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests.csproj"
+COPY "tests/NHSD.BuyingCatalogue.Organisations.Api.UnitTests/NHSD.BuyingCatalogue.Organisations.Api.UnitTests.csproj" "tests/NHSD.BuyingCatalogue.Organisations.Api.UnitTests/NHSD.BuyingCatalogue.Organisations.Api.UnitTests.csproj"
+COPY "tests/NHSD.BuyingCatalogue.Identity.Api.Testing.Data/NHSD.BuyingCatalogue.Identity.Api.Testing.Data.csproj" "tests/NHSD.BuyingCatalogue.Identity.Api.Testing.Data/NHSD.BuyingCatalogue.Identity.Api.Testing.Data.csproj"
```
[the remark this answers is not in the document]
Comment (sam-roberts-bjss) [reply samples/NHSD.BuyingCatalogue.Identity.Api.SampleMvcClient/Dockerfile]: The whole solution will not build if we do not copy the test projects as well.

File: samples/NHSD.BuyingCatalogue.Identity.Api.SampleResource/Dockerfile [hunk 10]
```diff
@@ -1,20 +1,31 @@
...
+RUN apt update && apt install iproute2 -y 
 
 FROM mcr.microsoft.com/dotnet/core/sdk:3.1-buster AS build
-WORKDIR /src
-COPY ["samples/NHSD.BuyingCatalogue.Identity.Api.SampleResource/NHSD.BuyingCatalogue.Identity.Api.SampleResource.csproj", "samples/NHSD.BuyingCatalogue.Identity.Api.SampleResource/"]
-RUN dotnet restore "samples/NHSD.BuyingCatalogue.Identity.Api.SampleResource/NHSD.BuyingCatalogue.Identity.Api.SampleResource.csproj"
+WORKDIR /
+COPY "NHSD.BuyingCatalogue.Identity-ExludeDatabase.sln" "NHSD.BuyingCatalogue.Identity-ExludeDatabase.sln"
```
[the remark this answers is not in the document]
Comment (sam-roberts-bjss) [reply samples/NHSD.BuyingCatalogue.Identity.Api.SampleResource/Dockerfile]: What are your concerns? 👍

File: samples/NHSD.BuyingCatalogue.Identity.Api.SampleMvcClient/Dockerfile (hunk 8 above)
[the remark this answers is not in the document]
Comment (assoomro) [reply samples/NHSD.BuyingCatalogue.Identity.Api.SampleMvcClient/Dockerfile]: I think we should add it in when we actually add a feature that uses it :]

File: samples/NHSD.BuyingCatalogue.Identity.Api.SampleMvcClient/Dockerfile (hunk 9 above)
[the remark this answers is not in the document]
Comment (assoomro) [reply samples/NHSD.BuyingCatalogue.Identity.Api.SampleMvcClient/Dockerfile]: Would it be possible to remove them from the solution then?

File: samples/NHSD.BuyingCatalogue.Identity.Api.SampleResource/Dockerfile (hunk 10 above)
[the remark this answers is not in the document]
Comment (assoomro) [reply samples/NHSD.BuyingCatalogue.Identity.Api.SampleResource/Dockerfile]: What concerns me is that now it seems like we have a dependency on other samples. Now we're bringing in all the samples for the sample specific build & have to copy all tests projects. Which doesn't seem right.

If we're building SampleMvcClient, we should bring in and build / restore only the things we need for this 'component'. In other words, in order to build a any subset of the system, we need a set which is much bigger, which could lead to belief that there are more dependencies across these subsystems than there are in reality.

**Pull request closed without merging**